File: risk_manager.py
"""
risk_manager.py

Enforces all position limits, drawdown rules, and Kelly sizing.
Every trade goes through this class before any order is placed.

Hard limits (non-negotiable):
  - Max 8% of portfolio per single trade
  - Halt trading at -20% daily loss
  - Kill bot permanently at -40% total drawdown
"""

import csv
import logging
import time
from datetime import date
from typing import TYPE_CHECKING

import config

if TYPE_CHECKING:
    from edge_detector import TradeSignal

logger = logging.getLogger(__name__)


class RiskManager:

    def __init__(self, starting_capital: float) -> None:
        self.starting_capital: float  = starting_capital
        self.portfolio_value: float   = starting_capital
        self.daily_start_value: float = starting_capital
        self._today: date             = date.today()
        self._killed: bool     = False
        self._halted_daily: bool = False
        self._open_positions: dict[str, float] = {}   # token_id -> USDC committed

        self._init_trade_log()

    # ── State ─────────────────────────────────────────────────────────────────

    @property
    def is_live(self) -> bool:
        """False if the bot should stop all trading activity."""
        return not self._killed and not self._halted_daily

    # ── Daily reset ───────────────────────────────────────────────────────────

    def _check_daily_reset(self) -> None:
        today = date.today()
        if today != self._today:
            logger.info(
                f"New day. Resetting daily PnL. Portfolio: ${self.portfolio_value:.2f}"
            )
            self.daily_start_value = self.portfolio_value
            self._today            = today
            self._halted_daily     = False

    # ── Limit checks ─────────────────────────────────────────────────────────

    def check_limits(self) -> bool:
        """
        Returns True if safe to place a trade.
        Call this before every order attempt.
        """
        self._check_daily_reset()

        if self._killed:
            logger.error("KILL SWITCH ACTIVE — bot is permanently stopped.")
            return False

        # ── Total drawdown ────────────────────────────────────────────────────
        total_dd = (self.starting_capital - self.portfolio_value) / self.starting_capital
        if total_dd >= config.TOTAL_DRAWDOWN_KILL_PCT:
            self._killed = True
            logger.critical(
                f"████ KILL SWITCH TRIGGERED ████ "
                f"Total drawdown: {total_dd:.1%} | "
                f"Portfolio: ${self.portfolio_value:.2f} | "
                f"Bot stopped permanently."
            )
            return False

        # ── Daily loss ────────────────────────────────────────────────────────
        if self.daily_start_value > 0:
            daily_dd = (self.daily_start_value - self.portfolio_value) / self.daily_start_value
            if daily_dd >= config.DAILY_LOSS_LIMIT_PCT:
                self._halted_daily = True
                logger.warning(
                    f"DAILY HALT | Daily loss: {daily_dd:.1%} | "
                    f"Resuming at next calendar day."
                )
                return False

        return True

    # ── Kelly sizing ──────────────────────────────────────────────────────────

    def kelly_size(self, win_prob: float, entry_price: float) -> float:
        """
        Fractional Kelly Criterion.

        entry_price: what we pay per share (0–1). Win payout = 1/entry_price - 1.
        Returns USDC position size, capped at MAX_POSITION_PCT of portfolio.
        """
        if not (0.01 < entry_price < 0.99):
            return 0.0

        b = (1.0 / entry_price) - 1.0    # net odds (profit per $1 risked)
        q = 1.0 - win_prob

        # Full Kelly fraction
        if b <= 0:
            return 0.0
        kelly_full = (b * win_prob - q) / b

        if kelly_full <= 0:
            return 0.0

        # Apply fractional Kelly + hard position cap
        fraction     = kelly_full * config.KELLY_FRACTION
        max_position = self.portfolio_value * config.MAX_POSITION_PCT
        size         = min(fraction * self.portfolio_value, max_position)

        return round(max(size, 0.0), 2)

    # ── Position tracking ─────────────────────────────────────────────────────

    def record_open(self, token_id: str, size_usdc: float) -> None:
        """Call immediately after a successful order placement."""
        self._open_positions[token_id] = (
            self._open_positions.get(token_id, 0.0) + size_usdc
        )
        logger.debug(f"Position opened: {token_id[:12]}… ${size_usdc:.2f}")

    def record_close(self, token_id: str, pnl: float, outcome: str, signal: "TradeSignal | None" = None) -> None:
        """
        Call when a position resolves (market settles on-chain).
        pnl: positive = profit, negative = loss (USDC)
        """
        _ = self._open_positions.pop(token_id, None)
        self.portfolio_value = round(self.portfolio_value + pnl, 4)

        emoji = "✅" if pnl > 0 else "❌"
        logger.info(
            f"{emoji} Closed | outcome={outcome} | pnl=${pnl:+.2f} | "
            f"portfolio=${self.portfolio_value:.2f}"
        )

        if signal:
            self._log_trade(signal, pnl, outcome)

        _ = self.check_limits()

    # ── Reporting ─────────────────────────────────────────────────────────────

    def summary(self) -> str:
        total_ret = (self.portfolio_value - self.starting_capital) / self.starting_capital
        daily_ret = (self.portfolio_value - self.daily_start_value) / self.daily_start_value
        return (
            f"Portfolio: ${self.portfolio_value:.2f} | "
            f"Total: {total_ret:+.2%} | "
            f"Daily: {daily_ret:+.2%} | "
            f"Open: {len(self._open_positions)} positions"
        )

    # ── CSV trade log ─────────────────────────────────────────────────────────

    def _init_trade_log(self) -> None:
        with open(config.TRADE_LOG_FILE, "w", newline="") as f:
            csv.writer(f).writerow([
                "timestamp", "market_id", "direction", "symbol",
                "price", "size_usdc", "edge", "p_model", "p_market",
                "momentum", "pnl", "portfolio_value", "outcome",
            ])

    def _log_trade(self, signal: "TradeSignal", pnl: float, outcome: str) -> None:
        with open(config.TRADE_LOG_FILE, "a", newline="") as f:
            csv.writer(f).writerow([
                time.strftime("%Y-%m-%d %H:%M:%S"),
                signal.market_id,
                signal.direction,
                signal.symbol,
                f"{signal.price:.4f}",
                f"{signal.size_usdc:.4f}",
                f"{signal.edge:.4f}",
                f"{signal.p_model:.4f}",
                f"{signal.p_market:.4f}",
                f"{signal.momentum:.6f}",
                f"{pnl:.4f}",
                f"{self.portfolio_value:.4f}",
                outcome,
            ])

File: src/risk_manager.rs
use std::collections::HashMap;
use std::fs::OpenOptions;
use std::io::Write;

use chrono::Local;
use tracing::{debug, error, info, warn};

use crate::config;
use crate::types::TradeSignal;

pub struct RiskManager {
    pub starting_capital: f64,
    pub portfolio_value: f64,
    pub daily_start_value: f64,
    today: chrono::NaiveDate,
    killed: bool,
    halted_daily: bool,
    open_positions: HashMap<String, f64>, // token_id -> USDC committed
}

impl RiskManager {
    pub fn new(starting_capital: f64) -> Self {
        let mut rm = Self {
            starting_capital,
            portfolio_value: starting_capital,
            daily_start_value: starting_capital,
            today: Local::now().date_naive(),
            killed: false,
            halted_daily: false,
            open_positions: HashMap::new(),
        };
        rm.init_trade_log();
        rm
    }

    // ── State ───────────────────────────────────────────────────────────────

    pub fn is_live(&self) -> bool {
        !self.killed && !self.halted_daily
    }

    // ── Daily reset ─────────────────────────────────────────────────────────

    fn check_daily_reset(&mut self) {
        let today = Local::now().date_naive();
        if today != self.today {
            info!(
                "New day. Resetting daily PnL. Portfolio: ${:.2}",
                self.portfolio_value
            );
            self.daily_start_value = self.portfolio_value;
            self.today = today;
            self.halted_daily = false;
        }
    }

    // ── Limit checks ────────────────────────────────────────────────────────

    pub fn check_limits(&mut self) -> bool {
        self.check_daily_reset();

        if self.killed {
            error!("KILL SWITCH ACTIVE — bot is permanently stopped.");
            return false;
        }

        // Total drawdown
        let total_dd =
            (self.starting_capital - self.portfolio_value) / self.starting_capital;
        if total_dd >= config::TOTAL_DRAWDOWN_KILL_PCT {
            self.killed = true;
            error!(
                "KILL SWITCH TRIGGERED | Total drawdown: {:.1}% | Portfolio: ${:.2} | Bot stopped permanently.",
                total_dd * 100.0,
                self.portfolio_value
            );
            return false;
        }

        // Daily loss
        if self.daily_start_value > 0.0 {
            let daily_dd =
                (self.daily_start_value - self.portfolio_value) / self.daily_start_value;
            if daily_dd >= config::DAILY_LOSS_LIMIT_PCT {
                self.halted_daily = true;
                warn!(
                    "DAILY HALT | Daily loss: {:.1}% | Resuming at next calendar day.",
                    daily_dd * 100.0
                );
                return false;
            }
        }

        true
    }

    // ── Kelly sizing ────────────────────────────────────────────────────────

    pub fn kelly_size(&self, win_prob: f64, entry_price: f64) -> f64 {
        if !(0.01 < entry_price && entry_price < 0.99) {
            return 0.0;
        }

        let b = (1.0 / entry_price) - 1.0; // net odds
        let q = 1.0 - win_prob;

        if b <= 0.0 {
            return 0.0;
        }
        let kelly_full = (b * win_prob - q) / b;

        if kelly_full <= 0.0 {
            return 0.0;
        }

        let fraction = kelly_full * config::KELLY_FRACTION;
        let max_position = self.portfolio_value * config::MAX_POSITION_PCT;
        let size = (fraction * self.portfolio_value).min(max_position);

        (size.max(0.0) * 100.0).round() / 100.0
    }

    // ── Position tracking ───────────────────────────────────────────────────

    pub fn record_open(&mut self, token_id: &str, size_usdc: f64) {
        *self.open_positions.entry(token_id.to_string()).or_insert(0.0) += size_usdc;
        debug!("Position opened: {}… ${:.2}", &token_id[..12.min(token_id.len())], size_usdc);
    }

    pub fn record_close(
        &mut self,
        token_id: &str,
        pnl: f64,
        outcome: &str,
        signal: Option<&TradeSignal>,
    ) {
        self.open_positions.remove(token_id);
        self.portfolio_value = ((self.portfolio_value + pnl) * 10000.0).round() / 10000.0;

        let emoji = if pnl > 0.0 { "+" } else { "-" };
        info!(
            "{emoji} Closed | outcome={outcome} | pnl=${pnl:+.2} | portfolio=${:.2}",
            self.portfolio_value
        );

        if let Some(sig) = signal {
            self.log_trade(sig, pnl, outcome);
        }

        self.check_limits();
    }

    // ── Reporting ───────────────────────────────────────────────────────────

    pub fn summary(&self) -> String {
        let total_ret =
            (self.portfolio_value - self.starting_capital) / self.starting_capital;
        let daily_ret =
            (self.portfolio_value - self.daily_start_value) / self.daily_start_value;
        format!(
            "Portfolio: ${:.2} | Total: {:+.2}% | Daily: {:+.2}% | Open: {} positions",
            self.portfolio_value,
            total_ret * 100.0,
            daily_ret * 100.0,
            self.open_positions.len()
        )
    }

    // ── CSV trade log ───────────────────────────────────────────────────────

    fn init_trade_log(&mut self) {
        if let Ok(mut f) = OpenOptions::new()
            .write(true)
            .create(true)
            .truncate(true)
            .open(config::TRADE_LOG_FILE)
        {
            let _ = writeln!(
                f,
                "timestamp,market_id,direction,symbol,price,size_usdc,edge,p_model,p_market,momentum,pnl,portfolio_value,outcome"
            );
        }
    }

    fn log_trade(&self, signal: &TradeSignal, pnl: f64, outcome: &str) {
        if let Ok(mut f) = OpenOptions::new()
            .append(true)
            .open(config::TRADE_LOG_FILE)
        {
            let ts = Local::now().format("%Y-%m-%d %H:%M:%S");
            let _ = writeln!(
                f,
                "{ts},{},{},{},{:.4},{:.4},{:.4},{:.4},{:.4},{:.6},{:.4},{:.4},{outcome}",
                signal.market_id,
                signal.direction,
                signal.symbol,
                signal.price,
                signal.size_usdc,
                signal.edge,
                signal.p_model,
                signal.p_market,
                signal.momentum,
                pnl,
                self.portfolio_value,
            );
        }
    }
}
